### PR TITLE
Feature: [P2P Distance] Enable P2P/splits in App

### PR DESCRIPTION
**Pull request by tienifr**
### Details

Enable P2P/Splits in App following [this plan](https://docs.google.com/document/d/15GQhipcqOJ69MbFA54KRUcWxjICURHYcfTlAb3eaVEQ/edit).

### Fixed Issues

$ https://github.com/Expensify/App/issues/36984
PROPOSAL: https://github.com/Expensify/App/issues/36984#issuecomment-1955950611

### Tests

1. Go to global create (`+`) > Request money
2. Change the request type to Distance
3. Choose and submit the start/finish points
4. Verify that the header title is ‘Distance’
5. Verify that you see both workspaces and individual users in the participants list
6. Verify that you can search for a user or a workspace
7. Verify that for each user/workspace the “Split” button is displayed
8. Choose one user and go to the confirmation page
9. Verify that the distance is displayed correctly (`Route pending...` for P2P request, or`73.45 miles @ $0.66 / mile` for workspace)
10. Verify that clicking the distance field navigates to the distance editing page
11. Change the waypoints >> Next
12. Verify the distance is modified accordingly
13. Go to any policy expense chat >> (`+`) >> Request money
14. Repeat step 2 - 4 and 9 - 12
15. Go to any group chat >> (`+`) >> Split bill
16. Verify user can create Distance split bill
17. Go to any DM >> (`+`) >> Request money
18. Verify user can create Distance split bill

- [x] Verify that no errors appear in the JS console

### Offline tests

NA

### QA Steps

1. Go to global create (`+`) > Request money
2. Change the request type to Distance
3. Choose and submit the start/finish points
4. Verify that the header title is ‘Distance’
5. Verify that you see both workspaces and individual users in the participants list
6. Verify that you can search for a user or a workspace
7. Verify that for each user/workspace the “Split” button is displayed
8. Choose one user and go to the confirmation page
9. Verify that the distance is displayed correctly (`Route pending...` for P2P request, or`73.45 miles @ $0.66 / mile` for workspace)
10. Verify that clicking the distance field navigates to the distance editing page
11. Change the waypoints >> Next
12. Verify the distance is modified accordingly
13. Go to any policy expense chat >> (`+`) >> Request money
14. Repeat step 2 - 4 and 9 - 12
15. Go to any group chat >> (`+`) >> Split bill
16. Verify user can create Distance split bill
17. Go to any DM >> (`+`) >> Request money
18. Verify user can create Distance split bill

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/113963320/781ddaa6-fec7-459c-a58f-39eeb7014ed6

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/113963320/532d8e85-248e-4bc1-90c7-cabedc4355a3

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/113963320/f82d735c-4b52-458f-9e4c-c85496ffd84b

https://github.com/Expensify/App/assets/113963320/1406df61-2cc9-4e11-8611-c1266d804520

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/113963320/30360348-2da0-4770-bfaf-cb3b070516e6

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/113963320/d18c5cd7-268f-48cc-9f1a-9a772d945ec2

https://github.com/Expensify/App/assets/113963320/6d8fcab9-8528-45d9-ad1c-c506dd8fd6e4

https://github.com/Expensify/App/assets/113963320/50b5438c-a72b-4a2a-aecf-3e6cb639edd2

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/113963320/d18c5cd7-268f-48cc-9f1a-9a772d945ec2

https://github.com/Expensify/App/assets/113963320/6d8fcab9-8528-45d9-ad1c-c506dd8fd6e4

https://github.com/Expensify/App/assets/113963320/50b5438c-a72b-4a2a-aecf-3e6cb639edd2

</details>

### CODE REVIEW

File: src/ONYXKEYS.ts
Comment (koko57) [inline src/ONYXKEYS.ts:130]: I know it's in the docs this way, but maybe it would be better to make it singular? (NVP_LAST_SELECTED_DISTANCE_RATE) @neil-marcellini wdyt?

Comment (neil-marcellini) [reply src/ONYXKEYS.ts:130]: I think plural makes sense because it's the set of all your last selected distance rates, per policy. When we use it though, we're mostly looking at one policy and one rate, so I understand your question.

Comment (neil-marcellini) [reply src/ONYXKEYS.ts:130]: I think it's fine as is.

File: src/libs/actions/IOU.ts
Comment (neil-marcellini) [inline src/libs/actions/IOU.ts:256]: Looks great, but needs to be updated to remove the customUnitRateID field name. The lastSelectedDistanceRates value is the customUnitID string directly, not an object.

File: src/libs/DistanceRequestUtils.ts
Comment (neil-marcellini) [inline src/libs/DistanceRequestUtils.ts:10]: Can we call this the customUnitRateID for consistency please?

File: src/libs/DistanceRequestUtils.ts
Comment (neil-marcellini) [inline src/libs/DistanceRequestUtils.ts:89]: Looks pretty good. @koko57, are we planning to use this function now when creating the merchant? I think so. If we don't use it now, when are we planning to start using it?

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looking pretty good, suggested some tweaks.

File: src/libs/DistanceRequestUtils.ts
Comment (koko57) [reply src/libs/DistanceRequestUtils.ts:89]: Yes, we can use it to get the merchant - we can refactor the existing function for getting the merchant to use this one and the one I've created while working on the new rate field - I will take care of this refactor in my PR

Comment (neil-marcellini) [reply src/libs/DistanceRequestUtils.ts:89]: Ok thanks. @tienifr I think we should keep this util, and also use it within the function that gets the merchant in this PR. Then @koko57 can add the piece for the rate in her PR.

**Comment by neil-marcellini** [comment]
> One thing to confirm: For group chat, do we allow distance split? @neil-marcellini

Yes, let's make that work too.

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.js
Comment (neil-marcellini) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.js:685]: I see that we're replacing the merchant with the distance only. Until we also add the rate field, let's continue to show the merchant which contains the distance and rate. Otherwise we wouldn't see the rate. Let's use getDistanceForDisplay within the getDistanceMerchant function, for now.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looking really good now, thanks for the updates! There's one thing I would like to see changed, and it sounds like we also need to enable this for group chats.

After that we can probably get this merged. The backend is almost done, but it will be a little while longer to finish up. Since we have the beta, I think it's ok if this goes out before the backend is done.

**Comment by neil-marcellini** [comment]
> I also think we should reflect this in design doc.

Sounds good. Please go ahead and make a suggestion there for how you would like to see it updated. I'll then review/accept it.

**Review by neil-marcellini — APPROVED** [review]
Good to go, pending C+ testing 🙂

**Comment by alitoshmatov** [comment]
Reviewing

**Comment by alitoshmatov** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/59907218/bc0630b6-2e1d-4c87-8684-653ce9fb86c0




</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/59907218/7b732d52-6f5a-4030-8990-66ed58c3f5b8





</details>

<details>
<summary>iOS: Native</summary>
https://github.com/Expensify/App/assets/59907218/3c5a5f45-d308-4518-9759-7f9d5b8a66dc





</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/59907218/d90ef770-88e7-4c49-b3c1-ceca5806ab5a





</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/59907218/33a844ed-f347-411f-9b74-af293c2fcad8




</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/59907218/1a4e972a-fba9-43e1-81e2-7b1cad41f9bb



</details>

**Review by alitoshmatov — APPROVED** [review]
Changes looks good to me

**Review by neil-marcellini — APPROVED** [review]
Mostly looks good, but I don't understand why we are showing "Route pending..." for the  distance field on P2P distance requests. It should display with the proper rate and distance text just like for a workspace.

Oh I see, it's because we need getRateForDisplay from this issue [[$500] [P2P Distance] Create a new Rate field](https://github.com/Expensify/App/issues/36985).

I'm fine if it doesn't fully work at this point, because it's under beta and we want to keep the PRs small 👍. I also can't think of any other reasonable way to do it, so looks good.

**Comment by neil-marcellini** [comment]
# ❗❗❗ Note to QA/bug reporters ❗❗❗

Please refrain from creating deploy blockers related to this PR/issue. The feature is under a beta that is only available to expensify accounts while we build the feature. It's expected to be incomplete right now.

**Pull request merged**
